### PR TITLE
Add shuffle sharding grouper/planner

### PR DESCRIPTION
**Pull request by ac1214**
Signed-off-by: Albert <ac1214@users.noreply.github.com>

**What this PR does**:

Implements generation of parallelize plans for the proposal outlined in https://github.com/cortexproject/cortex/pull/4272 using a shuffle sharding grouper and planner. Currently the parallelizable plans are generated but every compactor runs every planned compaction, the actual sharding will happen in a subsequent PR.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by ac1214** [comment]
I wanted to discuss a change in the block compaction behavior that this PR would introduce.

The current implementation of the Thanos compactor will always compact the first set of overlapping blocks if there exists such a set. This means that if the most recently ingested set of blocks from multiple ingesters are overlapping, the blocks will be compacted. If this happens, there is potentially a “missing” block in the compaction with the Thanos planner since there may be an ingester that has not fully uploaded the block when the compaction begins. So if there are 3 overlapping blocks when the compaction begins, and they are the latest blocks passed to the Thanos planner, the planner will plan a compaction of those 3 blocks even if there is a potential fourth ingester that has yet to upload a block.
	 
With this PR, overlapping blocks will not be compacted if they are the last set of blocks meaning in the example above, the 3 blocks won’t be compacted if there are the latest ones and they don’t cover a full range.
	 
In a real-world situation, this would only have an impact on customers who stop ingesting blocks. The impact will be that the last group of n blocks, where n is the number of ingesters will remain uncompacted for as long as they are the latest blocks. The impact of leaving the last n blocks uncompacted would be increased storage size as well as query time (if they continue to query even after stopping ingesting blocks). One thing to note with the Thanos approach, there can be duplicate work if the blocks are compacted, and another block that overlaps is uploaded after the compaction begins.
	 
A couple of different approaches I considered were adding grouping overlapping blocks before grouping by compactable ranges, this results in the compaction behavior being the same using these changes compared to Thanos. Another approach is if there are no new blocks after the time defined by the smallest block range passes from the max time of all the blocks, the block which are overlapping can be compacated, even if they are the latest blocks.
	 
Small example illustrating what’s mentioned above


4 total blocks
```
block 1: {
  MinTime: 0, MaxTime: 20
}
block 2: {
  MinTime: 21, MaxTime: 40
}
block 3: {
  MinTime: 21, MaxTime: 40
}
block 4: {
  MinTime: 21, MaxTime: 40
}
```

The above blocks with the current compaction with time ranges [2, 4, 8] would result in blocks:
```
block 1: {
    MinTime: 0, MaxTime: 20
}
block 2/3/4: {
MinTime: 21, MaxTime: 40
}
```
With this PR and the shuffle-sharding sharding strategy, the blocks would remain uncompacted.

**I was wondering what your thoughts were about which approach would be preferable?**

**Comment by ac1214** [comment]
This PR replaces and implements the changes recommended in https://github.com/cortexproject/cortex/pull/4318

**Comment by ac1214** [comment]
> I wanted to discuss a change in the block compaction behavior that this PR would introduce.
> 
> The current implementation of the Thanos compactor will always compact the first set of overlapping blocks if there exists such a set. This means that if the most recently ingested set of blocks from multiple ingesters are overlapping, the blocks will be compacted. If this happens, there is potentially a “missing” block in the compaction with the Thanos planner since there may be an ingester that has not fully uploaded the block when the compaction begins. So if there are 3 overlapping blocks when the compaction begins, and they are the latest blocks passed to the Thanos planner, the planner will plan a compaction of those 3 blocks even if there is a potential fourth ingester that has yet to upload a block.
> 
> With this PR, overlapping blocks will not be compacted if they are the last set of blocks meaning in the example above, the 3 blocks won’t be compacted if there are the latest ones and they don’t cover a full range.
> 
> In a real-world situation, this would only have an impact on customers who stop ingesting blocks. The impact will be that the last group of n blocks, where n is the number of ingesters will remain uncompacted for as long as they are the latest blocks. The impact of leaving the last n blocks uncompacted would be increased storage size as well as query time (if they continue to query even after stopping ingesting blocks). One thing to note with the Thanos approach, there can be duplicate work if the blocks are compacted, and another block that overlaps is uploaded after the compaction begins.
> 
> A couple of different approaches I considered were adding grouping overlapping blocks before grouping by compactable ranges, this results in the compaction behavior being the same using these changes compared to Thanos. Another approach is if there are no new blocks after the time defined by the smallest block range passes from the max time of all the blocks, the block which are overlapping can be compacted, even if they are the latest blocks.
> 
> Something else that I considered is making this a toggle to allow the user to define their own preference, but I think that this isn't ideal as it would either lead to having to support the toggle indefinitely or eventually having to have users switch to a single behavior.
> 
> Small example illustrating what’s mentioned above
> 
> 4 total blocks with 1 block incoming (not yet uploaded)
> 
> ```
> block 1: {
>   MinTime: 0, MaxTime: 20
> }
> block 2: {
>   MinTime: 21, MaxTime: 40
> }
> block 3: {
>   MinTime: 21, MaxTime: 40
> }
> block 4: {
>   MinTime: 21, MaxTime: 40
> }
> block 5 (incoming): {
>   MinTime: 21, MaxTime: 40
> }
> ```
> 
> **Thanos compaction**
> 
> The above blocks with the current (Thanos) compaction with time ranges [20, 120, 240] would result in blocks:
> 
> ```
> block 1: {
>   MinTime: 0, MaxTime: 20
> }
> block 2/3/4: {
>   MinTime: 21, MaxTime: 40
> }
> ```
> 
> Afterwards, once block 5 is fully uploaded the final resulting blocks from a single run of the compaction will be
> 
> ```
> block 1: {
>   MinTime: 0, MaxTime: 20
> }
> block 2/3/4: {
>   MinTime: 21, MaxTime: 40
> }
> block 5: {
>   MinTime: 21, MaxTime: 40
> }
> ```
> 
> With these blocks, another compaction will need to be done to fully compact the overlapping blocks 2-5.
> 
> **New compaction behavior**
> 
> With this PR and the shuffle-sharding strategy, the blocks would remain uncompacted. And would wait until a more recent block than 2-5 is uploaded. Once that block is uploaded blocks 2-5 would be impacted in 1 compaction.
> If there is a `block 6` uploaded with `MinTime: 41, MaxTime: 60` after `block 5` is fully uploaded, then the resulting blocks after a single compaction would be.
> 
> ```
> block 1: {
>   MinTime: 0, MaxTime: 20
> }
> block 2/3/4/5: {
>   MinTime: 21, MaxTime: 40
> }
> block 6: {
>   MinTime: 41, MaxTime: 60
> }
> ```
> 
> The downside with this approach is that the uncompacted blocks 2-5 were stored for a longer time compared to the current (Thanos) approach as it was waiting for a more recent block to be uploaded before compacting the blocks. In the above with this PR if `block 6` didn't exist, then blocks 2-5 would never be compacted as they would remain as the most recent blocks.
> 
> **I was wondering what your thoughts were about which approach would be preferable?**

Discussed in the community call and leaving the blocks uncompacted is okay.

**Comment by jeromeinsf** [comment]
Can you please also review this PR @alolita @Aneurysm9?

**Comment by jeromeinsf** [comment]
@pracucci and/or @bboreham how can we help with the review of this?

**Review by bboreham — APPROVED** [review]
The code is long and I didn't read through every line.  Broadly it looks ok.
I did wonder why the word "thanos" shows up so often - if the code is copied from Thanos it should say so, and if not can you just explain your thinking to me?

File: pkg/compactor/shuffle_sharding_grouper.go [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+		logger:                   logger,
+		bkt:                      bkt,
+		acceptMalformedIndex:     acceptMalformedIndex,
+		enableVerticalCompaction: enableVerticalCompaction,
+		reg:                      reg,
+		blocksMarkedForDeletion:  blocksMarkedForDeletion,
+		garbageCollectedBlocks:   garbageCollectedBlocks,
+		hashFunc:                 hashFunc,
+		compactions: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_compact_group_compactions_total",
```
[the remark this answers is not in the document]
Comment (ac1214) [reply pkg/compactor/shuffle_sharding_grouper.go:65]: Added a note where the metrics were copied from in Thanos. Even with these changes wouldn't the metrics in Cortex remain the same? They are only used when creating a new group using `compact.NewGroup` which is what is being done now.

**Comment by thejosephstevens** [comment]
Any way I can help with this PR? We're running into limits in our compaction (we have about 25M active time series in a single-tenant cortex). I'd be happy to run pre-release compactor builds if this needs some kind of validation.

File: pkg/compactor/shuffle_sharding_grouper.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/compactor/shuffle_sharding_grouper.go:65]: Then it might be better to expose those metrics as another function in Thanos?

**Review by alvinlin123 — APPROVED** [review]

**Comment by alvinlin123** [comment]
error message from build is 

```
go: updates to go.mod needed, disabled by -mod=vendor
	(Go version in go.mod is at least 1.14 and vendor directory exists.)
	to update it:
	go mod tidy
```

since this PR does not change `go.mod` I suspect the failure is due to that `go.mod` had change quite a bit since the submission of this PR. 

I will merge this PR, and keep a close eye on the automated workflow to see if things are ok.

**Comment by alvinlin123** [comment]
Please see https://github.com/cortexproject/cortex/pull/4621 instead.

**Pull request closed without merging**